upgrades/upgrade_testing_framework/core/docker_framework_client.py

```diff
@@ -0,0 +1,115 @@
+import logging
+import docker.client
+from docker.errors import ImageNotFound, NotFound
+from docker.models.networks import Network
+from docker.models.containers import Container
+from docker.models.volumes import Volume
+
+
+class RemoveRunningContainerException(Exception):
+    def __init__(self, container: Container):
+        self.container = container
+        super().__init__("A running container should be stopped before attempting to remove: {}".format(container.id))
+
+
+class ExistingContainerException(Exception):
+    def __init__(self, container_name: str):
+        self.container_name = container_name
+        super().__init__("There already exists a container with the name: {}".format(container_name))
+
+
+class DockerFrameworkClient:
+
+    def __init__(self, logger=logging.getLogger(__name__), docker_client=docker.from_env()):
+        # TODO: Use a custom logger
+        self.logger = logger
+        self.docker_client = docker_client
+
+    def is_container_running(self, container: Container) -> bool:
+        # TODO: Make more robust, case where this attr is not valid
+        if container is None:
+            raise TypeError("Provided argument is not of type docker.models.containers.Container")
+        container_state = container.attrs['State']
+        return container_state == 'running'
+
+    def does_container_exist(self, container_name: str) -> bool:
+        try:
+            self.docker_client.containers.get(container_id=container_name)
+        except NotFound:
+            return False
+        else:
+            return True
+
+    def does_image_exist_locally(self, image: str) -> bool:
```

Comment:
Same comment - this feels like it's an internal method rather than part of the modules external contract.  Recommend prefixing with an underscore.

Reply:
Agreed